Parallel.pm

```diff
@@ -0,0 +1,195 @@
+        if ( $childs_error->{$child_pid} > 0 ) {
+            $self->{RC} = 1;
+            if ( $debug ) { Log( "[DEBUG] SIG child error for $child_pid, rc=$CHILD_ERROR" ); }
+        }
+    };
+
+    eval {
+        local $SIG{ALRM} = sub { $self->{RC} = 1; croak 'alarm timeout'; };
+        alarm $self->{TIMEOUT};
+        $self->parallel_start();
+        alarm 0;
+    };
+
+    Log ( "[INFO] parallel run finished at " . ( strftime '%A, %d %B %Y %H:%M:%S GMT', localtime ) );
+    if ( $EVAL_ERROR && $EVAL_ERROR !~ quotemeta 'alarm timeout' ) {
+        $self->{RC} = 1; croak "[ERROR] $EVAL_ERROR";
+    } else {
+        foreach my $child_pid ( sort keys %$childs_error ) {
+            if ( $childs_error->{$child_pid} > 0 ) {
+                if ( $debug ) { Log( "[DEBUG] Child error for $child_pid, rc=" . $childs_error->{$child_pid} ); }
+                $self->{RC} = 1;
+                last;
+            }
+        }
+    }
+
+    return $self->{RC};
+}
+
+######################################################################
+# parallel_start
+#
+sub parallel_start {
+    my $self = shift;
+    my $num = $self->{COUNT};
+    my @counts = ( $COUNT_1..$num );
+    my $start_time = time();
+    my $end_time;
+    my $tout;
+    my $date;
+    my $children;
+    my $childs_error;
+
+    $children = $self->{CHILDREN};
+    $childs_error = $self->{CHILDS_ERROR};
+
+    foreach my $count (@counts) {
+        Log( "[${count}/$num] Process starting.." );
+        my $child_pid = fork;
+        if ( ! defined $child_pid ) { croak "[ERROR] Cannot fork: $ERRNO" };
+
+        if ( $child_pid == 0 ) { # child
+            $self->{PROC}( $count );
+        } else { # parent
+            $children->{$child_pid}     = $NUMBER_NEGATIVE_1;
+            $childs_error->{$child_pid} = $NUMBER_NEGATIVE_1;
+            $date = strftime '%A, %d %B %Y %H:%M:%S GMT', localtime;
+            Log( "[INFO] [$date] child started, proc=${count},pid=$child_pid");
+        }
+    }
+
+    while ( $self->parallel_children_number() ) {
+        foreach my $child_pid ( sort keys %$childs_error ) {
+            sleep $SLEEP_3;
+            $end_time = time();
+            $tout = $end_time - $start_time;
+            if ( $tout >= $self->{TIMEOUT} and defined $childs_error->{$child_pid} and $childs_error->{$child_pid} == $NUMBER_NEGATIVE_1 ) {
+                $date = strftime '%A, %d %B %Y %H:%M:%S GMT', localtime;
+                Log( "[INFO] [$date] kill ILL $child_pid" );
+                $childs_error->{$child_pid} = 1;
+                kill ILL => $child_pid;
+                if ( defined $children->{$child_pid} ) { delete $children->{$child_pid}; }
+                waitpid $child_pid, 0;
+            }
+        }
+    }
+
+    return 0;
+
+}
+
+######################################################################
+# parallel_children_number
+#   return number of running childs
+#
+sub parallel_children_number {
```

Comment:
Why do you need to call an external `ps` process just to track the number of spawned and still running children processes?